clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerLoginModule.java

```diff
@@ -256,22 +261,51 @@ public boolean login() throws LoginException {
             throw new IllegalStateException(String.format(
                     "Already have a committed token with private credential token count=%d; must login on another login context or logout here first before reusing the same login context",
                     committedTokenCount()));
-        OAuthBearerTokenCallback callback = new OAuthBearerTokenCallback();
+
+        attachToken();
+        attachExtensions();
+
+        log.info("Login succeeded; invoke commit() to commit it; current committed token count={}",
+                committedTokenCount());
+        return true;
+    }
+
+    private void attachToken() throws LoginException {
+        OAuthBearerTokenCallback tokenCallback = new OAuthBearerTokenCallback();
         try {
-            callbackHandler.handle(new Callback[] {callback});
+            callbackHandler.handle(new Callback[] {tokenCallback});
         } catch (IOException | UnsupportedCallbackException e) {
             log.error(e.getMessage(), e);
             throw new LoginException("An internal error occurred");
         }
-        tokenRequiringCommit = callback.token();
+
+        tokenRequiringCommit = tokenCallback.token();
         if (tokenRequiringCommit == null) {
-            log.info(String.format("Login failed: %s : %s (URI=%s)", callback.errorCode(), callback.errorDescription(),
-                    callback.errorUri()));
-            throw new LoginException(callback.errorDescription());
+            log.info(String.format("Login failed: %s : %s (URI=%s)", tokenCallback.errorCode(), tokenCallback.errorDescription(),
+                    tokenCallback.errorUri()));
+            throw new LoginException(tokenCallback.errorDescription());
         }
-        log.info("Login succeeded; invoke commit() to commit it; current committed token count={}",
-                committedTokenCount());
-        return true;
+    }
+
+    /*
+        Attaches SASL extensions to the Subject
+     */
+    private void attachExtensions() throws LoginException {
+        SaslExtensionsCallback extensionsCallback = new SaslExtensionsCallback();
+        try {
+            callbackHandler.handle(new Callback[] {extensionsCallback});
+        } catch (IOException e) {
+            log.error(e.getMessage(), e);
+            throw new LoginException("An internal error occurred");
+        } catch (UnsupportedCallbackException e) {
+            log.info("CallbackHandler " + callbackHandler.getClass().getName() + " does not support SASL extensions. No extensions will be added");
+        }
+        Map<String, String> unsecuredExtensions = extensionsCallback.extensions();
+        if (unsecuredExtensions ==  null) {
+            log.error("SASL Extensions cannot be null. Check whether your callback is explicitly setting them as null.");
+            throw new LoginException("Extensions cannot be null.");
+        }
+        subject.getPublicCredentials().add(extensionsCallback.extensions());
```

Comment:
That would be the best approach I think. I also found this way non-ideal but decided to stick with the implementation as with `ScramExtensions`

Reply:
@rondagostino @stanislavkozlovski Yes, it makes sense. We currently use `Map` for delegation tokens extension for SCRAM. Since we are only adding custom extensions to OAuth in this KIP, perhaps we should add `SaslExtensionsCallback` and `SaslExtensionsCallbackHandler` in `org.apache.kafka.common.security.auth` and use it only for `OAuth` for now. We can leave `ScramExtensionsCallback` and `ScramExtensionsCallbackHandler` as-is for now and update to use the common interface when we want to support customizable SCRAM extensions. At that point, we can deprecate `ScramExtensionsCallbackHandler` and use the new class. What do you think?

Reply:
Agreed, we can focus on OAUTHBEARER for now, take the right steps (adding 2 classes to the public API instead of just the Callback class), and start to leverage the added public API classes in other mechanisms (i.e. SCRAM-related) if/when the time seems right.  I am less familiar with the SCRAM-related code, so I defer to whatever/whenever you feel is best.

Reply:
Why should we not change SCRAM? Was the delegation token publicly accessible and that would break? I feel like we should change it outright if we can

Reply:
@stanislavkozlovski For this KIP, I think we should do one of these:
1. Implement custom extensions for OAUTHBEARER alone, but add common callback classes to enable reuse for other mechanisms in future. This means leaving SCRAM alone.
2. Support custom extensions for both SCRAM and OAUTHBEARER, changing the delegation token mechanism to use the custom extensions code path treating `token` as a property in the map.
I don't think we should do a half-change for SCRAM, changing a public contract without actually providing a good reason to do so (i.e. change the way extensions are propagated without supporting custom extensions).
Does that make sense?